channel-watcher.js

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var safeJSONParse = require('safe-json-parse/tuple');
+var fs = require('fs');
+var assert = require('assert');
+
+function ChannelWatcher(channel, opts) {
+    var self = this;
+
+    assert(channel.topChannel, 'must be a subChannel');
+    this.channel = channel;
+    this.filePath = opts.filePath;
+    this.logger = channel.logger;
+
+    this._boundReload = _boundReload;
+
+    this._validateFilePath();
+    this._establishFileWatcher();
+
+    // Trigger an initial load
+    this.reloadSync();
+
+    function _boundReload() {
+        self.reload();
+    }
+}
+
+ChannelWatcher.prototype._validateFilePath = function _validateFilePath() {
+    if (!fs.existsSync(this.filePath)) {
+        assert(false, 'Peer hosts file does not exist: ' + this.filePath);
+    }
+};
+
+ChannelWatcher.prototype._readPeerListFromFileSync =
+function _readPeerListFromFileSync() {
+    var tuple = safeJSONParse(fs.readFileSync(this.filePath, 'utf8'));
+    if (tuple[0]) {
+        assert(false, 'Invalid JSON for TChannel peer host file at ' + this.filePath);
```

Comment:
this could be dangerous,  in history,  we had outage because similar logic.   If the host-file update is not atomic,  i.e., not the file is updated chunk-by-chunk, and you will get events for a file that contains invalid JSON content,  this line might kill the whole node worker.  


@jwyuan and Sebastian Yates know the context.